### PR TITLE
The "Account Link" update

### PR DESCRIPTION
**Pull request by Awzaw**
This breaks getPlayer(IPlayer $player) : array by changing it to getPlayerData(string $playername) mainly because sending the Player object is impractical when modifying the Player itself. Also because anything called getPlayer() should... get a player, not an array.

MySql, yaml and sqlite are all fully working, including antihack and account linking/unlinking.
The dreaded "you don't have permission to login" and register is also fixed by adapting some code from SimpleAuthHelper - which is still compatible.

~~Would only be merged if we get access to Player::$username in PlayerPreLoginEvent.~~

### CODE REVIEW

**Comment by SOF3** [comment]
For Player::setName(), would it be better to use reflections?

**Comment by SOF3** [comment]
Also, what about plugins that assume Player->getName() returns a consistent value? Many other references won't be updated.

**Comment by SOF3** [comment]
Also consider using DataPacketReceiveEvent to change the username.

**Comment by Awzaw** [comment]
I don't think we'll get a setName()  function, but Player::$username could be made accessible via PlayerPreLoginEvent, from what I understand - where could it use reflection, and how? As for plugins expecting a consistent value, I don't think that's a problem as the change happens early  enough during authentication and  I'm prompting the user to relog  instead of changing $username on the fly - I also tried using transfer and kick, but both have severe drawbacks. Authentication by 'lastip' (UUID) also works well with account linking.  I'll look into using DataPacketReceiveEvent.

**Comment by SOF3** [comment]
If other plugins already use the name before SimpleAuth does, it will still have a problem.

File: src/SimpleAuth/EventListener.php [hunk 1]
```diff
@@ -80,28 +82,37 @@ public function onPlayerPreLogin(PlayerPreLoginEvent $event){
...
+    /**
+     * @param DataPacketReceiveEvent $event
+     *
+     * @priority HIGHEST
+     */
+
+    public function onDataPacketReceive(DataPacketReceiveEvent $event){
+        if($event->getPacket() instanceof LoginPacket){
+            if(!$this->plugin->getConfig()->get("allowLinking")){
+                return true;
```
Comment (SOF3) [inline src/SimpleAuth/EventListener.php]: Why?

**Review by SOF3 — CHANGES_REQUESTED** [review]

File: src/SimpleAuth/EventListener.php (hunk 1 above)
Comment (Awzaw) [reply src/SimpleAuth/EventListener.php]: fixed, thanks

**Pull request closed without merging**